flytepropeller/pkg/controller/nodes/array/handler.go

```diff
@@ -631,6 +651,99 @@ func (a *arrayNodeHandler) Handle(ctx context.Context, nCtx interfaces.NodeExecu
+		return err
+	}
+
+	// update the literal with the offloaded URI
+	if err := createZeroValueRecursively(toBeOffloaded); err != nil {
+		logger.Errorf(ctx, "Failed to update the literal with offloaded URI [%s] with error [%s]", dataReference, err)
+		return err
+	}
+
+	toBeOffloaded.Uri = dataReference.String()
+	toBeOffloaded.SizeBytes = literalSizeBytes
+	return nil
+}
+
+// createZeroValueRecursively creates a zero value for the literal recursively
+func createZeroValueRecursively(toBeOffloaded *idlcore.Literal) error {
+	if toBeOffloaded.GetScalar() != nil {
+		switch toBeOffloaded.GetScalar().Value.(type) {
+		case *idlcore.Scalar_Primitive:
+			switch toBeOffloaded.GetScalar().GetPrimitive().Value.(type) {
+			case *idlcore.Primitive_Integer:
+				toBeOffloaded.GetScalar().GetPrimitive().Value = &idlcore.Primitive_Integer{Integer: 0}
+			case *idlcore.Primitive_FloatValue:
+				toBeOffloaded.GetScalar().GetPrimitive().Value = &idlcore.Primitive_FloatValue{FloatValue: 0}
+			case *idlcore.Primitive_StringValue:
+				toBeOffloaded.GetScalar().GetPrimitive().Value = &idlcore.Primitive_StringValue{StringValue: ""}
+			case *idlcore.Primitive_Boolean:
+				toBeOffloaded.GetScalar().GetPrimitive().Value = &idlcore.Primitive_Boolean{Boolean: false}
+			case *idlcore.Primitive_Datetime:
+				toBeOffloaded.GetScalar().GetPrimitive().Value = &idlcore.Primitive_Datetime{Datetime: &timestamp.Timestamp{Seconds: 0}}
+			case *idlcore.Primitive_Duration:
+				toBeOffloaded.GetScalar().GetPrimitive().Value = &idlcore.Primitive_Duration{Duration: &duration.Duration{Seconds: 0}}
+			default:
+				return fmt.Errorf("unsupported primitive type [%T]", toBeOffloaded.GetScalar().GetPrimitive().GetValue())
+			}
+		case *idlcore.Scalar_Generic:
+			toBeOffloaded.GetScalar().Value = &idlcore.Scalar_Generic{Generic: &structpb.Struct{}}
+		case *idlcore.Scalar_Blob:
+			toBeOffloaded.GetScalar().Value = &idlcore.Scalar_Blob{Blob: &idlcore.Blob{}}
+		case *idlcore.Scalar_Error:
+			toBeOffloaded.GetScalar().Value = &idlcore.Scalar_Error{Error: &idlcore.Error{}}
+		case *idlcore.Scalar_NoneType:
+			toBeOffloaded.GetScalar().Value = &idlcore.Scalar_NoneType{NoneType: &idlcore.Void{}}
+		case *idlcore.Scalar_Binary:
+			toBeOffloaded.GetScalar().Value = &idlcore.Scalar_Binary{Binary: &idlcore.Binary{
+				Value: []byte{},
+			}}
+		case *idlcore.Scalar_StructuredDataset:
+			toBeOffloaded.GetScalar().Value = &idlcore.Scalar_StructuredDataset{StructuredDataset: &idlcore.StructuredDataset{}}
+		case *idlcore.Scalar_Schema:
+			toBeOffloaded.GetScalar().Value = &idlcore.Scalar_Schema{Schema: &idlcore.Schema{}}
+		case *idlcore.Scalar_Union:
+			toBeOffloaded.GetScalar().Value = &idlcore.Scalar_Union{Union: &idlcore.Union{}}
+		default:
+			return fmt.Errorf("unsupported scalar type [%T]", toBeOffloaded.GetScalar().GetValue())
+		}
+	} else if toBeOffloaded.GetCollection() != nil {
+		for _, l := range toBeOffloaded.GetCollection().Literals {
```

Comment:
If we proceed with this approach, we should do this once (for the first element of the literals then bail... While Literals support heterogeneous collections, LiteralType doesn't and therefore we don't throughout the system.

Reply:
I would prefer to go with explicit LiteralType which i find it better too. Discard this change